src/java.base/share/classes/sun/security/util/Pem.java

```diff
@@ -25,30 +25,304 @@
+
+        String data = sb.toString();
+
+        // Verify footer starts with 5 hyphens.
+        do {
+            switch (is.read()) {
+                case '-' -> hyphen++;
+                case -1 -> throw new IllegalArgumentException(
+                    "Input ended prematurely");
+                default -> throw new IllegalArgumentException(
+                    "Incomplete footer");
+            }
+        } while (hyphen < 5);
+
+        hyphen = 0;
+        sb = new StringBuilder(64);
+        sb.append("-----");
+
+        // Look for Complete header by looking for the end of the hyphens
+        do {
+            switch (c = is.read()) {
+                case '-' -> hyphen++;
+                case -1 -> throw new IllegalArgumentException(
+                    "Input ended prematurely");
+                default -> sb.append((char) c);
+            }
+        } while (hyphen == 0);
+
+        // Verify ending with 5 hyphens.
+        do {
+            switch (is.read()) {
+                case '-' -> hyphen++;
+                case -1 -> throw new IllegalArgumentException(
+                    "Input ended prematurely");
+                default -> throw new IllegalArgumentException(
+                    "Incomplete footer");
+            }
+        } while (hyphen < 5);
+
+        while ((c = is.read()) != eol && c != -1 && c != '\r' && c != WS) {
+            throw new IllegalArgumentException("Invalid PEM format:  " +
+                "No EOL char found in footer:  0x" +
+                HexFormat.of().toHexDigits((byte) c));
+        }
+
+        sb.append("-----");
+        String footer = sb.toString();
+        if (footer.length() < 14 || !footer.startsWith("-----END ") ||
+            !footer.endsWith("-----")) {
+            throw new IllegalArgumentException("Illegal footer: " + footer);
+        }
+
+        // Verify the object type in the header and the footer are the same.
+        String headerType = header.substring(11, header.length() - 5);
+        String footerType = footer.substring(9, footer.length() - 5);
+        if (!headerType.equals(footerType)) {
+            throw new IllegalArgumentException("Header and footer do not " +
+                "match: " + headerType + " " + footerType);
+        }
+
+        // If there was data before finding the 5 dashes of the PEM header,
+        // backup 5 characters and save that data.
+        byte[] preData = null;
+        if (os.size() > 5) {
+            preData = Arrays.copyOf(os.toByteArray(), os.size() - 5);
         }
+
+        return new PEMRecord(
+            header.substring(11, header.lastIndexOf('-') - 4), data,
+            preData);
     }
+
+    public static PEMRecord readPEM(InputStream is) throws IOException {
+        return readPEM(is, false);
+    }
+
+    /**
+     * Construct a String-based encoding based off the type.
+     * @return the string
+     */
+    public static String pemEncoded(String type, byte[] data) {
```

Comment:
You could refactor the common code in these two methods into a separate method and call that with the type and data.